primitives/src/lib.rs

```diff
@@ -63,3 +63,60 @@ pub mod metrics {
 		}
 	}
 }
+
+#[cfg(test)]
+mod test {
+	use super::metrics::*;
+	use substrate_prometheus_endpoint::Registry;
+
+	#[test]
+	fn metrics_works_when_starting_without_registry() {
+		let metrics = MiningMetrics::new(None).unwrap();
+		let initial_count = metrics.count();
+		assert_eq!(initial_count, 0);
+
+		// increase
+		metrics.inc();
+		assert_eq!(metrics.count(), initial_count + 1);
+
+		// add
+		metrics.add(4);
+		assert_eq!(metrics.count(), initial_count + 5);
+
+		// elapsed time
+		assert!(metrics.elapsed().as_nanos() > 0);
+	}
+
+	#[test]
+	fn metrics_works_when_starting_with_registry() {
+		let registry = Registry::new();
+		let metrics = MiningMetrics::new(Some(&registry)).unwrap();
+		let initial_count = metrics.count();
+		assert_eq!(initial_count, 0);
+
+		// increase
+		metrics.inc();
+		assert_eq!(metrics.count(), initial_count + 1);
+
+		// add
+		metrics.add(4);
+		assert_eq!(metrics.count(), initial_count + 5);
+
+		// elapsed time
+		assert!(metrics.elapsed().as_nanos() > 0);
+
+		// there's only 1 metric counter registered in MiningMetricsInner::register()
+		let results = registry.gather();
+		assert_eq!(results.len(), 1);
+	}
+
+	#[test]
+	#[allow(clippy::redundant_clone)]
```

Comment:
Don't you find `Drop` more correct instead? Clones can usually be 'shallow' copies and have special interactions with smart pointer types. Dropping may seem more robust at first sight, it may not matter in this case. Even if the above doesn't matter, the explicit drop is more informative about the test's intent than an allow attribute, imo.

Reply:
I'm not sure what you mean here. The allow attribute is here just to silence the clippy warning.

Reply:
I mean `drop(metrics)`